src/auth-strategies/azure_ad_strategy.ts

```diff
@@ -59,12 +59,19 @@ export const setupBearerStrategy = (
       (
         _: express.Request,
         profile: AdUser,
-        done: (err: Error | undefined, profile?: AdUser) => void
+        done: (
+          err: Error | undefined,
+          user?: AdUser & { readonly kind: "azure-ad" }
```

Comment:
Maybe we cab define a new type with the kind of strategy and an enumeration with the available strategy we use to extract a user?
```suggestion
          user?: UserData
```
a possible implementation:
```ts
export enum AvailableStrategyEnum {
  AZURE,
  OTHER
}
export type AvailableStrategyEnumType = t.TypeOf<
  typeof AvailableStrategyEnumType
>;
export const AvailableStrategyEnumType = enumType<AvailableStrategyEnum>(
  AvailableStrategyEnum,
  "AvailableStrategyEnumType"
);
export const UserAuthData = t.intersection([
  AdUser,
  t.interface({ kind: AvailableStrategyEnumType })
]);
export type UserAuthData = t.TypeOf<typeof UserAuthData>;
```

Reply:
My idea is we already have `SessionUser`, which maps `user data + strategy type`. Ideally, `SessionUser` will be defined as:
```ts
type SessionUser = 
| AdUser & { readonly kind: "azure-ad" }
| OtherKindUser & { readonly kind: "other" }
| AnotherKindUser & { readonly kind: "another" }
```

in your example, how would you use `UserAuthData`? Would it be specific for each kind, or would it be a generic type? In each case, to be honest, I see no much value in adding an intermediate type.

Nevertheless, your question made me doubt if we really need this `kind` attribute. In the end, once a token  is resolved using the appropriate strategy, and received user data are in valid shape, we have no need to know where data came from.

Reply:
Allright, your reply and example clear all my doubts. Anyway, your thought about the really need on a `kind` attribute is right because the valid shape will tell us where data came from as you said.

Reply:
I did it in https://github.com/pagopa/io-developer-portal-backend/pull/144/commits/5566b7e3a4d54ef73c92fcfc3efd2a58f06601a0

Anyway, if we will need it, we can easily re-introduce the `kind` attribute